### PR TITLE
Respect selection sort order and fields in attribute table copy

### PR DESCRIPTION
**Pull request by NathanW2**
PR to allow maintaining the order of the selection as well as the fields that have been selected for that view.

Before this PR the attribute table would use the selection order that is in the layer selection list, which may or may not follow the order in the table itself.  Best to respect the sort here I think in order to give the user back what they see in the UI.

Limiting the fields also better control over the data that you copy to the clipboard.  

image

image

WIth some more testing I would like to get this into 3.4 as it's improves the UX quite a lot for the attribute table.

### CODE REVIEW

**Comment by NathanW2** [comment]
@nirvn feelings on this one?

**Comment by nirvn** [comment]
@NathanW2 , nice, it's something that I've long felt should be addressed too.

Is this change, https://github.com/qgis/QGIS/pull/8048/files#diff-566e6f75f165cf8af3620ab161209984R47, an API break? If so that'd have to be reverted.

**Comment by NathanW2** [comment]
@nirvn yeah it was. Fixed now.

**Comment by NathanW2** [comment]
A nice side effect of this now is that you can copy and paste just a given set of fields from the table into a new layer and it just takes those fields.

Pasted memory layer is on the bottom:

image

**Comment by NathanW2** [comment]
If I fix those failed tests.  What is the feelings for this in 3.4? @nyalldawson  @m-kuhn @wonder-sk etc

**Comment by nirvn** [comment]
@NathanW2 , I'm +1 to merge now, this is as much a fix as it is an improvement.

File: src/app/qgsattributetabledialog.cpp [hunk 1]
```diff
@@ -762,7 +765,46 @@ void QgsAttributeTableDialog::mActionCutSelectedRows_triggered()
...
-  QgisApp::instance()->copySelectionToClipboard( mLayer );
+  // TODO: Loop selected rows
+  if ( mMainView->view() == QgsDualView::AttributeTable )
+  {
+    QList<QgsFeatureId> featureIds = mMainView->tableView()->selectedFeaturesIds();
+    QgsFeatureStore featureStore;
+    QgsFields fields = QgsFields( mLayer->fields() );
+    QStringList fieldNames;
+
+    Q_FOREACH ( const QgsAttributeTableConfig::ColumnConfig &columnConfig, mMainView->attributeTableConfig().columns() )
```
Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: Someone decided Q_FOREACH is no longer cool, so now the cool code is more lines:

```
const auto configs = mMainView->attributeTableConfig().columns();
for ( const QgsAttributeTableConfig::ColumnConfig &columnConfig : configs )
```

File: src/app/qgsattributetabledialog.cpp [hunk 2]
```diff
@@ -762,7 +765,46 @@ void QgsAttributeTableDialog::mActionCutSelectedRows_triggered()
 
 void QgsAttributeTableDialog::mActionCopySelectedRows_triggered()
 {
-  QgisApp::instance()->copySelectionToClipboard( mLayer );
+  // TODO: Loop selected rows
+  if ( mMainView->view() == QgsDualView::AttributeTable )
+  {
+    QList<QgsFeatureId> featureIds = mMainView->tableView()->selectedFeaturesIds();
```
Comment (m-kuhn) [inline src/app/qgsattributetabledialog.cpp]: const (because after you use `for( fid : featureIds ) `

File: src/auth/oauth2/qgsauthoauth2config.cpp [hunk 3]
```diff
@@ -11,6 +11,7 @@
  *   (at your option) any later version.                                   *
  *                                                                         *
  ***************************************************************************/
+#include <functional>
```
Comment (m-kuhn) [inline src/auth/oauth2/qgsauthoauth2config.cpp]: Is this related?

File: src/gui/attributetable/qgsattributetableview.cpp [hunk 4]
```diff
@@ -104,6 +104,27 @@ void QgsAttributeTableView::setAttributeTableConfig( const QgsAttributeTableConf
     }
     i++;
   }
+  mConfig = config;
+}
+
+QList<QgsFeatureId> QgsAttributeTableView::selectedFeaturesIds()
+{
+  QgsFeatureIds featureIds = mFeatureSelectionManager->selectedFeatureIds();
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetableview.cpp]: const

File: src/gui/attributetable/qgsattributetableview.cpp [hunk 5]
```diff
@@ -104,6 +104,27 @@ void QgsAttributeTableView::setAttributeTableConfig( const QgsAttributeTableConf
...
+  QModelIndexList indexList;
+  Q_FOREACH ( const QgsFeatureId &id, featureIds )
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetableview.cpp]: `for ( id: featureIds)`

File: src/gui/attributetable/qgsattributetableview.cpp [hunk 6]
```diff
@@ -104,6 +104,27 @@ void QgsAttributeTableView::setAttributeTableConfig( const QgsAttributeTableConf
...
+  {
+    QModelIndex index = mFilterModel->fidToIndex( id );
+    indexList << index;
+  }
+
+  qSort( indexList );
+  QList<QgsFeatureId> ids;
+  Q_FOREACH ( const QModelIndex &index, indexList )
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetableview.cpp]: .

File: src/gui/attributetable/qgsattributetableview.h [hunk 7]
```diff
@@ -80,6 +82,13 @@ class GUI_EXPORT QgsAttributeTableView : public QTableView
      */
     void setAttributeTableConfig( const QgsAttributeTableConfig &config );
 
+    /**
+     * Returns the selected features in the attribute table in sorted order.
+     * \returns The selected features in the attribute table.
+     * \since QGIS 3.4
+     */
+    QList<QgsFeatureId> selectedFeaturesIds();
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetableview.h]: `QList<QgsFeatureId> selectedFeatureIds() const;`

File: src/gui/attributetable/qgsorganizetablecolumnsdialog.cpp [hunk 8]
```diff
@@ -94,15 +97,15 @@ QgsOrganizeTableColumnsDialog::QgsOrganizeTableColumnsDialog( const QgsVectorLay
...
+}
 
-  QgsSettings settings;
-  restoreGeometry( settings.value( QStringLiteral( "Windows/QgsOrganizeTableColumnsDialog/geometry" ) ).toByteArray() );
+QgsOrganizeTableColumnsDialog::QgsOrganizeTableColumnsDialog( const QgsVectorLayer *vl, QWidget *parent, Qt::WindowFlags flags )
+  : QgsOrganizeTableColumnsDialog( vl, vl->attributeTableConfig(), parent, flags )
+{
 }
 
 QgsOrganizeTableColumnsDialog::~QgsOrganizeTableColumnsDialog()
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsorganizetablecolumnsdialog.cpp]: `~QgsOrganizeTableColumnsDialog() = default` in the header

Comment (m-kuhn) [reply src/gui/attributetable/qgsorganizetablecolumnsdialog.cpp]: or kill it?

File: src/gui/attributetable/qgsdualview.h [hunk 9]
```diff
@@ -198,6 +198,8 @@ class GUI_EXPORT QgsDualView : public QStackedWidget, private Ui::QgsDualViewBas
      */
     QString sortExpression() const;
 
+    QgsAttributeTableConfig attributeTableConfig() { return mConfig; }
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsdualview.h]: const

And for extra points move implemenation to cpp

**Comment by m-kuhn** [comment]
I think this brings QGIS to "expected behaviour", so it's a UX fix.

**Comment by NathanW2** [comment]
Ok thanks.  I will review the feedback and get all the tests passing for
merge.

File: src/gui/attributetable/qgsattributetableview.h [hunk 10]
```diff
@@ -80,6 +82,13 @@ class GUI_EXPORT QgsAttributeTableView : public QTableView
      */
     void setAttributeTableConfig( const QgsAttributeTableConfig &config );
 
+    /**
+     * Returns the selected features in the attribute table in sorted order.
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetableview.h]: I'd just expand on this a bit - sorted order is a bit vague. E.g.

   Returns the selected features in the attribute table, in the order that they are shown in the table (i.e. respecting the current attribute table column sorting preferences).

**Comment by nyalldawson** [comment]
SOrry - just realised I had a pending review sitting here which I hadn't clicked submit for!

**Comment by NathanW2** [comment]
@m-kuhn @nyalldawson thanks for the feedback. Addressing them now.

File: src/gui/attributetable/qgsattributetableview.cpp [hunk 11]
```diff
@@ -104,6 +104,30 @@ void QgsAttributeTableView::setAttributeTableConfig( const QgsAttributeTableConf
...
+{
+  QgsFeatureIds featureIds = mFeatureSelectionManager->selectedFeatureIds();
+  QModelIndexList indexList;
+  Q_FOREACH ( const QgsFeatureId &id, featureIds )
+  {
+    QModelIndex index = mFilterModel->fidToIndex( id );
+    indexList << index;
+  }
+
+  qSort( indexList );
```
[the remark this answers is not in the document]
Comment (NathanW2) [reply src/gui/attributetable/qgsattributetableview.cpp]: std::sort doesn't seem to build when used here.

[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/gui/attributetable/qgsattributetableview.cpp]: You need to use:

    std::sort( indexList.begin(), indexList.end() )

[the remark this answers is not in the document]
Comment (NathanW2) [reply src/gui/attributetable/qgsattributetableview.cpp]: Yep worked it out just after I posted it.

**Comment by NathanW2** [comment]
Thanks for the feedback. Merged into master now for some more testing.
